### ISSUE TITLE
Blocklist dysfunctional `morx` table

### ISSUE THREAD

**Issue by khaledhosny** [harfbuzz/harfbuzz#4108]
Currently we use some heuristics to decide whether to use AAT or OT layout table when fonts provide both and our heuristics try to match what CoreText does. All good. However, some fonts have AAT tables for historical reasons (since OSX initially didn’t support OT layout) and on non-mac platforms people were getting OT layout, now they get AAT and it can be unexpected (sometimes the AAT table are even there by accident, FontForge sometimes ”translates” OT to AAT and does it badly).

So, I’m asking for a way to always prefer OT. Currently our `ot` shaper does both OT and AAT layout, so changing shaper list order does not help.

One way, though probably will be ugly internally, is to introduce and new dummy `aat` shaper that simply prefers AAT, and may be make `ot` do the reverse and add a new `default` (or some other name) that is equivalent to current `ot` shaper (which is misnomer already).

Or may be some other way, I don’t really mind, though shaper list is an already existing mechanism and some clients already expose it to users (I know at least luaotfload and SILE).

**Comment by behdad** [comment]
Changing `ot` to disable `aat` is probably fine. We can add a new default `hb` to do what `ot` currently does. What should a default `aat` do though? Only apply `morx` and never `GSUB`? What about the positioning part? There are fonts with `morx` and `GPOS`.

**Comment by khaledhosny** [comment]
I’d think it would prefer AAT tables, but I’m not sure if it should be all or none (i.e. if `morx` is available only apply `kern` and skip `GPOS`), but I guess it makes more sense since it is likely to be used mainly for testing and debugging purposes.

**Comment by khaledhosny** [comment]
Here is another case where this is needed. This seems to be an Moroccan font, it contains non-functional `morx` table (seems to have been autogenerated with FontForge, probably by mistake), but since Windows does not support AAT shaping it went unnoticed.

[AALMAGHRIBI.ttf.zip](https://github.com/harfbuzz/harfbuzz/files/13171764/AALMAGHRIBI.ttf.zip)

**Comment by khaledhosny** [comment]
Regardless of the issue here, I don’t know if we can find a heuristic to detect such broken `morx` tables, though I know nothing about how `morx` works or in what aspect this font’s table is broken.

**Comment by khaledhosny** [comment]
Reading this issue again, I probably opened it precisely about this font since this is not the first time I encounter it, though I can’t remember or find where I saw it before.

**Comment by dscorbett** [comment]
After reviewing AALMAGHRIBI’s 'morx' table, I think there is a bug in HarfBuzz’s AAT implementation.

This font does not enable its topographical subtables’ subfeatures in its `defaultFlags`. They are only enabled when the `kCursiveConnection` feature is set to `kCursiveSelector` (which it should always be for Arabic). However, `feature_mappings` does not have a mapping for that feature–value pair. Therefore, HarfBuzz never enables the relevant subfeatures.

**Comment by khaledhosny** [comment]
> After reviewing AALMAGHRIBI’s 'morx' table, I think there is a bug in HarfBuzz’s AAT implementation.
> 
> This font does not enable its topographical subtables’ subfeatures in its `defaultFlags`. They are only enabled when the `kCursiveConnection` feature is set to `kCursiveSelector` (which it should always be for Arabic). However, `feature_mappings` does not have a mapping for that feature–value pair. Therefore, HarfBuzz never enables the relevant subfeatures.

I don’t know anything about `morx`, but the font is broken with Core Text as well, which why I thought it is a font issue.

**Comment by jfkthame** [comment]
According to the Apple font feature registry at https://developer.apple.com/fonts/TrueType-Reference-Manual/RM09/AppendixF.html:

**Cursive Connection**

Feature Constant | Feature Value | Feature Type
-- | -- | --
kCursiveConnection | 2 | Exclusive

The Cursive Connection feature type is used for cursively-connected scripts. It is required for Arabic, but may be used for other scripts as well.

Selector Name | Selector Constant(s) | Selector Value | Selector Function
-- | -- | -- | --
Unconnected | kUnconnectedSelector | 0 | Disables cursive connection altogether. Note that selecting this for some scripts will result in incorrect liguistic appearance.
Partially Connected | kPartiallyConnectedSelector | 1 | Selects pre-drawn letterforms that connect in a non-contextual manner.
Cursive | kCursiveSelector | 2 | Selects full, contextual connection of letterforms. _**This setting must always be the default in fonts for scripts that require this behavior.**_

(emphasis added).

So if the font fails to set this in its `defaultFlags`, I think we can reasonably dismiss it as a font bug. (Especially if it fails similarly in Core Text.)

**Comment by dscorbett** [comment]
You’re right: I just tried the font in Pages, and the cursive connection does work but only when explicitly enabled. So a heuristic could be that if an Arabic font’s 'morx' doesn’t enable cursive by default, HarfBuzz should fall back to OpenType.

**Comment by khaledhosny** [comment]
I don’t need this anymore, if someone else needs it in the future, it will be their problem to handle.

**Comment by khaledhosny** [comment]
This came up again in https://github.com/flutter/flutter/issues/160841, the font seems a bit popular and seems to be also unlikely to get fixed (it was released by a Moroccan official body whose website is now defunct, https://mushafmohammedi.com/), so I’m wondering if we should do something like GDEF blocklist and block the `morx` table of this font?

**Comment by behdad** [comment]
> This came up again in [flutter/flutter#160841](https://github.com/flutter/flutter/issues/160841), the font seems to be a bit popular and also seems unlikely to get fixed (it was released by a Moroccan official body whose website is now defunct, https://mushafmohammedi.com/), so I’m wondering if we should do something like GDEF blocklist and block the `morx` table of this font?

Works for me. Though, maybe we should match on morx & gsub & gdef sizes to be more confident.
